PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc

```diff
@@ -5,168 +5,167 @@
+      }
+
+      // set output node names; will get all outputs if `output_names` is not provided
+      std::vector<const char*> run_output_node_names;
+      if (output_names.empty()) {
+        run_output_node_names = output_node_names_;
+      } else {
+        for (const auto& name : output_names) {
+          run_output_node_names.push_back(name.c_str());
+        }
+      }
+
+      // run
+      auto output_tensors = session_->Run(RunOptions{nullptr},
+                                          input_node_names_.data(),
+                                          input_tensors.data(),
+                                          input_tensors.size(),
+                                          run_output_node_names.data(),
+                                          run_output_node_names.size());
+
+      // convert output to floats
+      FloatArrays outputs;
+      for (auto& output_tensor : output_tensors) {
+        assert(output_tensor.IsTensor());
+
+        // get output shape
+        auto tensor_info = output_tensor.GetTensorTypeAndShapeInfo();
+        auto length = tensor_info.GetElementCount();
+
+        auto floatarr = output_tensor.GetTensorMutableData<float>();
+        outputs.emplace_back(floatarr, floatarr + length);
+      }
+      assert(outputs.size() == run_output_node_names.size());
+
+      return outputs;
     }
-    assert(outputs.size() == run_output_node_names.size());
 
-    return outputs;
-  }
+    const std::vector<std::string>& ONNXRuntime::getOutputNames() const {
+      if (session_) {
+        return output_node_strings_;
+      } else {
+        throw cms::Exception("RuntimeError") << "Needs to call createSession() first before getting the output names!";
+      }
+    }
 
-  const std::vector<std::string>& ONNXRuntime::getOutputNames() const {
-    if (session_) {
-      return output_node_strings_;
-    } else {
-      throw cms::Exception("RuntimeError") << "Needs to call createSession() first before getting the output names!";
+    const std::vector<int64_t>& ONNXRuntime::getOutputShape(const std::string& output_name) const {
+      auto iter = output_node_dims_.find(output_name);
+      if (iter == output_node_dims_.end()) {
+        throw cms::Exception("RuntimeError") << "Output name " << output_name << " is invalid!";
+      } else {
+        return iter->second;
+      }
     }
-  }
-
-  const std::vector<int64_t>& ONNXRuntime::getOutputShape(const std::string& output_name) const {
-    auto iter = output_node_dims_.find(output_name);
-    if (iter == output_node_dims_.end()) {
-      throw cms::Exception("RuntimeError") << "Output name " << output_name << " is invalid!";
-    } else {
-      return iter->second;
+
+    Env& ONNXRuntime::getEnv() {
+      // `env` cannot be const as the constructor of Ort::Session takes Env&.
+      // But the implementation of Env is thread-safe so it's fine.
+      CMS_THREAD_SAFE static Env env(ORT_LOGGING_LEVEL_WARNING, "onnxruntime");
+      return env;
```

Comment:
did the other option, to make this const and return by value fail?

Reply:
looking at https://github.com/cms-sw/cmsdist/pull/5346#issuecomment-553863363, it appears that my comment was unclear.

I meant that here we have:
```c++
Env ONNXRuntime::getEnv() {
static const Env env(ORT_LOGGING_LEVEL_WARNING, "onnxruntime");
return env;
```

If I understood correctly, ```Env``` is essentially a shared-like pointer to the larger environment blob and a copy should be low in cost.

Reply:
@slava77 Returning by value would not work here I think, as `Env` has deleted copy and assignment constructors (https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L87-L89). 
I think a cleaner fix is to make the `Session` constructor takes a `const Env&` in the C++ API (in the C API it actually takes a `const OrtEnv *` so I do not know why it's not made const in the C++ API). Then we can simply use `const Env&` in our code: https://github.com/hqucms/cmssw/blob/onnxruntime-master-const-static/PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc#L22.

Reply:
Since https://github.com/cms-sw/cmsdist/pull/5346 has been updated I pushed the commit https://github.com/cms-sw/cmssw/pull/28112/commits/7f10c2a3d3ed06a1c9321001e739d179d8a0043b with the corresponding changes in CMSSW.

Reply:
Since https://github.com/cms-sw/cmsdist/pull/5346 has been updated I pushed the commit https://github.com/cms-sw/cmssw/pull/28112/commits/7f10c2a3d3ed06a1c9321001e739d179d8a0043b with the corresponding changes in CMSSW.